contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java

```diff
@@ -0,0 +1,434 @@
+import java.util.Optional;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.Op;
+import org.apache.zookeeper.OpResult;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.Watcher.Event.KeeperState;
+import org.apache.zookeeper.ZooDefs;
+import org.apache.zookeeper.ZooKeeper;
+import org.apache.zookeeper.data.ACL;
+import org.apache.zookeeper.data.Stat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A thin wrapper around vanilla zookeeper client to facilitate the usage of
+ * <a href="www.github.com/linkedin/parseq">Parseq</a>.
+ *
+ * @author Ang Xu
+ */
+public class ZKClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ZKClient.class);
+
+  private volatile ZooKeeper _zkClient;
+  private volatile Reaper _reaper;
+
+  private final String _connectionString;
+  private final int _sessionTimeout;
+  private final Engine _engine;
+
+  private final Watcher _defaultWatcher = new DefaultWatcher();
+  private final Object _mutex = new Object();
+  private final StateListener _listener = new StateListener();
+  private KeeperState _currentState = null;
+
+  /**
+   * Constructs a {@link ZKClient} with the given parameters.
+   *
+   * @param connectionString comma separated host:port pairs, each corresponding to
+   *                         a zk server.
+   * @param sessionTimeout   session timeout in milliseconds.
+   * @param engine           a parseq {@link Engine} to schedule background tasks.
+   */
+  public ZKClient(String connectionString, int sessionTimeout, Engine engine) {
+    _connectionString = connectionString;
+    _sessionTimeout = sessionTimeout;
+    _engine = engine;
+  }
+
+  /**
+   * Starts the zookeeper connection. The returned promise will be
+   * resolved once the connection is established.
+   *
+   * @return promise
+   */
+  public Promise<Void> start() {
+    AtomicBoolean committed = new AtomicBoolean(false);
+    SettablePromise<Void> promise = Promises.settable();
+    _listener.subscribe(KeeperState.SyncConnected,
+        (Promise p) -> {
+          if (committed.compareAndSet(false, true)) {
+            if (p.isFailed()) {
+              promise.fail(p.getError());
+            } else {
+              promise.done(null);
+            }
+          }
+        }
+    );
+
+    try {
+      _zkClient = new ZooKeeper(_connectionString, _sessionTimeout, _defaultWatcher);
+      _reaper = new Reaper(_engine);
+    } catch (IOException ex) {
+      if (committed.compareAndSet(false, true)) {
+        promise.fail(ex);
+      }
+    }
+    return promise;
+  }
+
+  /**
+   * Shuts down the zookeeper connection.
+   * @throws InterruptedException
+   */
+  public void shutdown() throws InterruptedException {
```

Comment:
But it's a blocking call anyway unless I submit the shutdown tasks to another executor, which seems overkilled.